File: labs/lab-01-n-plus-one-profiling/typescript/solution.ts
// Lab 01: N+1 Query Profiling — TypeScript reference solution (PostgreSQL).
//
//   Run:  node --experimental-strip-types solution.ts
import { createRequire } from 'node:module';
import assert from 'node:assert';

const require = createRequire(import.meta.url);
const { db } = require('../../../tooling/node/labkit');

const USER_COUNT = 100;
const POSTS_PER_USER = 5;

interface Post { id: number; title: string; }
interface UserPosts { id: number; name: string; posts: Post[]; }

async function setupDataset(): Promise<void> {
  await db.exec('DROP TABLE IF EXISTS n1_posts');
  await db.exec('DROP TABLE IF EXISTS n1_users');
  await db.exec('CREATE TABLE n1_users (id INT PRIMARY KEY, name TEXT NOT NULL, email TEXT NOT NULL)');
  await db.exec('CREATE TABLE n1_posts (id INT PRIMARY KEY, title TEXT NOT NULL, body TEXT NOT NULL, user_id INT NOT NULL REFERENCES n1_users(id))');
  await db.exec('CREATE INDEX idx_n1_posts_user_id ON n1_posts(user_id)');
  await db.exec("INSERT INTO n1_users SELECT g, 'User ' || g, 'user' || g || '@example.com' FROM generate_series(1, $1) g", [USER_COUNT]);
  await db.exec("INSERT INTO n1_posts SELECT u*10+p, 'Post ' || p || ' by User ' || u, 'body', u FROM generate_series(1, $1) u, generate_series(0, $2) p", [USER_COUNT, POSTS_PER_USER - 1]);
  db.resetCounters();
}

async function fetchNPlusOne(): Promise<UserPosts[]> {
  const users = await db.query('SELECT id, name FROM n1_users ORDER BY id');
  const out: UserPosts[] = [];
  for (const u of users) {
    const posts = await db.query('SELECT id, title FROM n1_posts WHERE user_id = $1', [u.id]);
    out.push({ id: u.id, name: u.name, posts });
  }
  return out;
}

async function fetchJoin(): Promise<UserPosts[]> {
  const rows = await db.query(
    'SELECT u.id AS user_id, u.name AS user_name, p.id AS post_id, p.title AS post_title ' +
    'FROM n1_users u LEFT JOIN n1_posts p ON p.user_id = u.id ORDER BY u.id, p.id'
  );
  const byId = new Map<number, UserPosts>();
  for (const r of rows) {
    if (!byId.has(r.user_id)) byId.set(r.user_id, { id: r.user_id, name: r.user_name, posts: [] });
    if (r.post_id != null) byId.get(r.user_id)!.posts.push({ id: r.post_id, title: r.post_title });
  }
  return [...byId.values()];
}

async function fetchInBatch(): Promise<UserPosts[]> {
  const users = await db.query('SELECT id, name FROM n1_users ORDER BY id');
  const ids = users.map((u: any) => u.id);
  const posts = await db.query('SELECT id, title, user_id FROM n1_posts WHERE user_id = ANY($1)', [ids]);
  const byUser = new Map<number, Post[]>(users.map((u: any) => [u.id, []]));
  for (const p of posts) byUser.get(p.user_id)!.push({ id: p.id, title: p.title });
  return users.map((u: any) => ({ id: u.id, name: u.name, posts: byUser.get(u.id)! }));
}

class PostDataLoader {
  private queue: number[] = [];
  private futures = new Map<number, (posts: Post[]) => void>();
  private scheduled = false;

  load(userId: number): Promise<Post[]> {
    return new Promise<Post[]>((resolve) => {
      this.futures.set(userId, resolve);
      this.queue.push(userId);
      if (!this.scheduled) { this.scheduled = true; queueMicrotask(() => this._dispatch()); }
    });
  }

  async _dispatch(): Promise<void> {
    const ids = [...new Set(this.queue)];
    this.queue = [];
    this.scheduled = false;
    const rows = await db.query('SELECT id, title, user_id FROM n1_posts WHERE user_id = ANY($1)', [ids]);
    const byUser = new Map<number, Post[]>();
    for (const r of rows) {
      if (!byUser.has(r.user_id)) byUser.set(r.user_id, []);
      byUser.get(r.user_id)!.push({ id: r.id, title: r.title });
    }
    for (const [uid, resolve] of this.futures) resolve(byUser.get(uid) || []);
    this.futures.clear();
  }
}

async function fetchWithDataloader(): Promise<UserPosts[]> {
  const users = await db.query('SELECT id, name FROM n1_users ORDER BY id');
  const loader = new PostDataLoader();
  const posts = await Promise.all(users.map((u: any) => loader.load(u.id)));
  return users.map((u: any, i: number) => ({ id: u.id, name: u.name, posts: posts[i] }));
}

const countPosts = (rows: UserPosts[]): number => rows.reduce((n, r) => n + r.posts.length, 0);

async function main(): Promise<void> {
  await setupDataset();
  console.log(`Seeded ${USER_COUNT} users x ${POSTS_PER_USER} posts in Postgres`);

  db.resetCounters();
  const n1 = await fetchNPlusOne();
  assert.strictEqual(db.queryCount, 101, 'N+1 should run 101 queries');

  db.resetCounters();
  const join = await fetchJoin();
  assert.strictEqual(db.queryCount, 1, 'JOIN should run 1 query');

  db.resetCounters();
  const batch = await fetchInBatch();
  assert.strictEqual(db.queryCount, 2, 'IN batch should run 2 queries');

  db.resetCounters();
  const dl = await fetchWithDataloader();
  assert.strictEqual(db.queryCount, 2, 'DataLoader should run 2 queries (1 users + 1 batched)');

  for (const r of [n1, join, batch, dl]) {
    assert.strictEqual(r.length, 100, 'all return 100 users');
    assert.strictEqual(countPosts(r), 500, 'all return 500 posts');
  }

  console.log('OK — N+1=101, JOIN=1, IN batch=2, DataLoader=2 queries');
  await db.close();
}

main().catch((e: Error) => { console.error(e.message); process.exit(1); });

File: labs/lab-01-n-plus-one-profiling/python/solution.py
#!/usr/bin/env python3
"""
Lab 01: N+1 Query Profiling — reference solution (PostgreSQL).

Runs against real Postgres through labkit, so the query count maps to real
database round trips — exactly where N+1 hurts. `db.query_count` is the
instrument: it increments on every query, so you can prove 101 -> 1 -> 2.

The lab creates and seeds its own tables (n1_users, n1_posts); it does not
touch the shared `users` table used by other labs.

Run:   python solution.py
Test:  python -m unittest test_lab.py
"""

import asyncio
from collections import defaultdict

from labkit import db

USER_COUNT = 100
POSTS_PER_USER = 5


# ──────────────────────────────────────────────
# SETUP — create + seed this lab's own tables (given)
# ──────────────────────────────────────────────

def setup_dataset():
    db.execute("DROP TABLE IF EXISTS n1_posts")
    db.execute("DROP TABLE IF EXISTS n1_users")
    db.execute(
        "CREATE TABLE n1_users (id INT PRIMARY KEY, name TEXT NOT NULL, email TEXT NOT NULL)"
    )
    db.execute(
        "CREATE TABLE n1_posts ("
        " id INT PRIMARY KEY, title TEXT NOT NULL, body TEXT NOT NULL,"
        " user_id INT NOT NULL REFERENCES n1_users(id))"
    )
    db.execute("CREATE INDEX idx_n1_posts_user_id ON n1_posts(user_id)")
    db.execute(
        "INSERT INTO n1_users "
        "SELECT g, 'User ' || g, 'user' || g || '@example.com' "
        "FROM generate_series(1, %s) g",
        (USER_COUNT,),
    )
    db.execute(
        "INSERT INTO n1_posts "
        "SELECT u * 10 + p, 'Post ' || p || ' by User ' || u, 'body of post', u "
        "FROM generate_series(1, %s) u, generate_series(0, %s) p",
        (USER_COUNT, POSTS_PER_USER - 1),
    )
    db.reset_counters()


# ──────────────────────────────────────────────
# STEP 1 — Naive N+1 (the baseline you are fixing)
# ──────────────────────────────────────────────

def fetch_users_with_posts_n_plus_one() -> list[dict]:
    """1 query for users + N queries for each user's posts = N+1 round trips."""
    users = db.query("SELECT id, name FROM n1_users ORDER BY id")
    result = []
    for user in users:
        posts = db.query(
            "SELECT id, title FROM n1_posts WHERE user_id = %s",
            (user["id"],),
        )
        result.append({"id": user["id"], "name": user["name"], "posts": posts})
    return result


# ──────────────────────────────────────────────
# STEP 2 — Fix #1: JOIN (1 query)
# ──────────────────────────────────────────────

def fetch_users_with_posts_join() -> list[dict]:
    rows = db.query(
        "SELECT u.id AS user_id, u.name AS user_name, p.id AS post_id, p.title AS post_title "
        "FROM n1_users u LEFT JOIN n1_posts p ON p.user_id = u.id "
        "ORDER BY u.id, p.id"
    )
    users: dict[int, dict] = {}
    for row in rows:
        uid = row["user_id"]
        if uid not in users:
            users[uid] = {"id": uid, "name": row["user_name"], "posts": []}
        if row["post_id"] is not None:
            users[uid]["posts"].append({"id": row["post_id"], "title": row["post_title"]})
    return list(users.values())


# ──────────────────────────────────────────────
# STEP 3 — Fix #2: IN batch via = ANY (2 queries)
# ──────────────────────────────────────────────

def fetch_users_with_posts_in_batch() -> list[dict]:
    users = db.query("SELECT id, name FROM n1_users ORDER BY id")
    user_ids = [u["id"] for u in users]
    posts = db.query(
        "SELECT id, title, user_id FROM n1_posts WHERE user_id = ANY(%s)",
        (user_ids,),
    )
    posts_by_user: dict[int, list] = {u["id"]: [] for u in users}
    for post in posts:
        posts_by_user[post["user_id"]].append({"id": post["id"], "title": post["title"]})
    return [
        {"id": u["id"], "name": u["name"], "posts": posts_by_user[u["id"]]}
        for u in users
    ]


# ──────────────────────────────────────────────
# PART 2 — DataLoader (batch N async loads into 1 query)
# ──────────────────────────────────────────────

class PostDataLoader:
    """Collect per-user lookups within one tick, then fire ONE = ANY query."""

    def __init__(self):
        self._queue: list[int] = []
        self._futures: dict[int, asyncio.Future] = {}
        self._scheduled = False

    async def load(self, user_id: int) -> list[dict]:
        loop = asyncio.get_event_loop()
        future = loop.create_future()
        self._futures[user_id] = future
        self._queue.append(user_id)
        if not self._scheduled:
            self._scheduled = True
            loop.call_soon(lambda: asyncio.ensure_future(self._dispatch()))
        return await future

    async def _dispatch(self):
        user_ids = list(set(self._queue))
        self._queue.clear()
        self._scheduled = False
        rows = db.query(
            "SELECT id, title, user_id FROM n1_posts WHERE user_id = ANY(%s)",
            (user_ids,),
        )
        posts_by_user: dict[int, list] = defaultdict(list)
        for row in rows:
            posts_by_user[row["user_id"]].append({"id": row["id"], "title": row["title"]})
        for user_id, future in self._futures.items():
            if not future.done():
                future.set_result(posts_by_user.get(user_id, []))
        self._futures.clear()


async def fetch_users_with_posts_dataloader() -> list[dict]:
    """1 query for users + 1 batched query for posts, despite N load() calls."""
    users = db.query("SELECT id, name FROM n1_users ORDER BY id")
    loader = PostDataLoader()
    posts_lists = await asyncio.gather(*[loader.load(u["id"]) for u in users])
    return [
        {"id": u["id"], "name": u["name"], "posts": posts}
        for u, posts in zip(users, posts_lists)
    ]


def main():
    print("=" * 60)
    print("LAB 01: N+1 Query Profiling (PostgreSQL)")
    print("=" * 60)
    setup_dataset()
    print(f"Seeded {USER_COUNT} users x {POSTS_PER_USER} posts in Postgres\n")

    for name, fn in [
        ("N+1 (naive)", fetch_users_with_posts_n_plus_one),
        ("JOIN (eager)", fetch_users_with_posts_join),
        ("IN batch (ANY)", fetch_users_with_posts_in_batch),
    ]:
        db.reset_counters()
        rows = fn()
        print(f"{name:<16} -> {len(rows)} users in {db.query_count} queries")

    db.reset_counters()
    dl = asyncio.run(fetch_users_with_posts_dataloader())
    print(f"{'DataLoader':<16} -> {len(dl)} users in {db.query_count} queries "
          f"(1 users + 1 batched), despite {len(dl)} load() calls")


if __name__ == "__main__":
    main()
